CWE number as enum

darkspirit510 opened the pull request:
Migrated CweNumber to enum to ensure there is no magic number stuff going on. Could remove a lot of mapping code that just attempts to map string to number. Since the code has to go through CweNumber's lookup method (which informs about request of unknown number), this (at least somehow) handles https://github.com/OWASP-Benchmark/BenchmarkUtils/issues/6 and https://github.com/OWASP-Benchmark/BenchmarkUtils/issues/7. It obviously does not handle default blocks for mapping of any tools internal namings/ids to CweNumbers.

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/CweNumber.java:
```diff
@@ -1,165 +1,530 @@
- * @author Sascha Knoop
- * @created 2021
- */
 package org.owasp.benchmarkutils.score;
```
Can you add the header you removed from this file back in?

darkspirit510 replied:
Done

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/AcunetixReader.java:
```diff
@@ -230,40 +227,12 @@ private TestCaseResult parseAcunetixReportItem(Node flaw) throws Exception {
-        // Add any 'new' CWEs ever found to switch above so we know they are mapped properly.
-        System.out.println("INFO: Found following CWE which we haven't seen before: " + cweNum);
-        return Integer.parseInt(cweNum);
+        return CweNumber.lookup(cweNum);
```
This change doesn't allow us to manually decide how to map these. This can be especially important for DAST tools as they frequently report 1 type of injection as another. I'm thinking in cases like this we should leave the manual mapping the way it was.  What do you think?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/BurpReader.java:
```diff
@@ -131,45 +131,45 @@ static int cweLookup(String id) {
+                return CweNumber
+                        .SENSITIVE_INFORMATION_IN_BROWSER_CACHE; // Information Exposure Through
+                // Browser Caching-Cacheable HTTPS
+                // Response
```
Can you merge the above 2 lines into a single line comment?

darkspirit510 replied:
Already fixed

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/CASTAIPReader.java:
```diff
@@ -99,47 +98,11 @@ private TestCaseResult parseCASTAIPIssue(Node flaw) throws Exception {
-        }
-        return 0000;
+
+        return CweNumber.lookup(name.trim());
```
Again, I think its better to leave the manual mapping here so we can map in anything 'new'.

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/CheckmarxESReader.java:
```diff
@@ -116,22 +111,23 @@ private boolean isIrrelevant(String name) {
                 return CweNumber.WEAK_RANDOM;
         }
-        return cwe;
+
```
I think we should add a warning like the others, e.g., "something new seen here not mapped before". Take exact message from other parser that already has this warning.

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/CheckmarxReader.java:
```diff
@@ -217,17 +217,18 @@ private TestCaseResult parseCheckmarxVulnerability(Node query, Node result) {
                 return CweNumber.WEAK_RANDOM;
         }
-        return cwe;
+
```
Can you add new/missing mapping warning here?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/CodeQLReader.java:
```diff
@@ -190,75 +190,11 @@ private TestCaseResult parseLGTMFinding(
-                                + " for rule: "
-                                + ruleName);
+    private CweNumber mapCWE(Integer cweNumber) {
+        if (cweNumber == 335) {
```
We should retain the explanatory comment of why we are doing this.

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/CodeQLReader.java:
```diff
@@ -190,75 +190,11 @@ private TestCaseResult parseLGTMFinding(
         }
-        return 0; // Not mapped to anything
+
+        return CweNumber.lookup(cweNumber);
```
I prefer that we keep the manual mapping so we can get the warning of 'new' types of issues not yet mapped.

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/InsiderReader.java:
```diff
@@ -79,8 +79,8 @@ private TestCaseResult parseTestCaseResult(JSONObject finding) {
-                int cwe = cweNumber(finding);
-                tcr.setCWE(cwe);
+                String cwe = finding.getString("cwe").substring(4);
+                tcr.setCWE(CweNumber.lookup(cwe));
```
Restore what we had so we can flag reported CWEs not yet mapped. And there is one mapping here that doesn't go straight up, 326 to Weak Crypto.

darkspirit510 replied:
Done

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/NJSScanReader.java:
```diff
@@ -213,34 +213,16 @@ private TestCaseResult produceTestCaseResult(JSONObject file, int cwe_identifier
         }
 
-        return cwe;
+        return CweNumber.lookup(cwe);
```
Can we restore what we had so we can flag reported CWEs not yet mapped?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SemgrepReader.java:
```diff
@@ -54,34 +54,23 @@ public TestSuiteResults parse(ResultFile resultFile) throws Exception {
-                                + cwe);
         }
-        return cwe;
+        return CweNumber.lookup(cwe);
```
Can we restore what we had so we can flag reported CWEs not yet mapped?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ShiftLeftReader.java:
```diff
@@ -67,30 +68,30 @@ public TestSuiteResults parse(ResultFile resultFile) throws Exception {
-                return 79;
+                return CweNumber.XSS;
             default:
                 throw new RuntimeException("Unknown category: " + category);
```
Maybe change this to a simple system out println? Runtime exception seems harsh.

darkspirit510 replied:
I can change this, but I didn't touch this in my PR.

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnappyTickReader.java:
```diff
@@ -115,40 +116,17 @@ private int extractTestNumber(String testfile) {
+                return CweNumber.PATH_TRAVERSAL;
         }
-        return 0;
+
```
Can we restore what we had so we can flag reported CWEs not yet mapped?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SonarQubeReader.java:
```diff
@@ -210,347 +210,277 @@ public static int cweLookup(String squidNumber) {
+                return CweNumber.DONTCARE;
+            case "S2681": // Multiline blocks should be enclosed in curly braces
+                return CweNumber.INCORRECT_BLOCK_DELIMITATION;
             case "S2696":
```
For these 3, can you move the comments up to match similar changes throughout?

darkspirit510 replied:
Done

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/VisualCodeGrepperReader.java:
```diff
@@ -161,7 +161,8 @@ private int figureCWE(TestCaseResult tcr, Node catnode) {
             default:
-                return 00; // System.out.println( "Unknown vuln category for VisualCodeGrepper: " +
+                return CweNumber.DONTCARE; // System.out.println( "Unknown vuln category for
+                // VisualCodeGrepper: " +
```
Please merge these 2 comment lines together.

darkspirit510 replied:
Done

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/WapitiReader.java:
```diff
@@ -106,47 +106,17 @@ public TestSuiteResults parse(ResultFile resultFile) throws Exception {
             if (title.startsWith("CWE-")) {
                 String cweNum = title.substring("CWE-".length(), title.indexOf(":"));
-                cwe = cweLookup(cweNum);
+                cwe = CweNumber.lookup(cweNum);
```
Can we restore the old cweLookup so we can still get the warning "WARNING: Wapiti-Unmapped CWE number: " + cwe);"?

darkspirit510 replied:
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/ZapJsonReader.java:
```diff
@@ -152,55 +152,16 @@ private String extractTestName(String fullUrl) {
+        if ("264".equals(cwe)) {
+            return CweNumber.IMPROPER_ACCESS_CONTROL;
         }
+
```
Can we restore the old cweLookup so we can still get the warning "WARNING: ZAP CWE not mapped to expected test suite CWE: " + cwe);

davewichers commented on plugin/src/test/java/org/owasp/benchmarkutils/score/CweNumberTest.java:
```diff
@@ -0,0 +1,74 @@
+package org.owasp.benchmarkutils.score;
```
Do we have preambles/copyright notices for test files? I think we do. Can you add one here?

darkspirit510 replied:
Done

davewichers commented:
@darkspirit510 - Thanks so much for all this refactoring work! I provided some minor comments in a few dozen of the files changed. Can you address those comments and we should be good to go.

darkspirit510 commented:
Sorry for the late reply :-) Since most of your comments are about the "missing" default message, one general reply why I removed them:

After checking all mapping methods in all Reader classes, you can divide them into three groups:

1. Mapping from tool-internal string/code/whatever to CWE number. This can definitly not be automated and is untouched from me (except, of course, the enum conversion)
2. Mapping from one CWE number to another. This is "hacky", because Benchmark has a limited set of expected CWE numbers, although there are some more detailed CWE numbers describing the same thing (e. g. CWE-89 "SQL injection" and CWE-564 "Hibernate SQL Injection"). Since Benchmark does not have any fuzzyness/forgiveness and only accepts a specific CWE numbers, some readers contain mappings to "fix"/map one CWE number to another. This is sad, but has to stay in the reader.
3. The only type of mappings I removed (which for some classes means remove the mapping code completly) is mapping one CWE number (sometimes as String) to exactly the same CWE number. If a given number did not exist in a mapping method, there was a log message. What's the value of this, if it is just mapping a number on itself? If a tool already reports CWE numbers, why not just pass them to the enum and only warn if the enum does not (yet) contain this number.

davewichers commented:
As you address issues per my comments above, can you add comments to each issue you've addressed so I can resolve comments on fixed issues?

darkspirit510 commented:
If you accept my explaination, i can write a short note to every comment.

davewichers commented:
@darkspirit510 - that works. Thanks.

darkspirit510 commented on plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/HorusecReader.java:
```diff
@@ -102,37 +102,17 @@ private int figureCwe(JSONObject vuln) {
-                System.out.println("WARN: Horusec reported CWE not yet mapped: " + cwe);
-                return Integer.parseInt(cwe);
         }
+
```
As discussed: https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379

davewichers commented:
@darkspirit510 - Did you ever work on my comments (per above)? Also, there is a branch conflict now, w/SemgrepReader. Can you resolve that too?

darkspirit510 commented:
@davewichers most of the comments should be explained in comment https://github.com/OWASP-Benchmark/BenchmarkUtils/pull/31#issuecomment-1304863379. But I actually missed some comments, I'll go through them within the next days.

darkspirit510 commented:
I fixed/commented all the things. If you're still not sure about this PR, I think about closing and redoing it in smaller chunks.

davewichers commented:
@darkspirit510 - There was a lot of work done here, but might have been overcome by events. Do you want finish this one? Or start over and do this in smaller chunks?

davewichers commented:
> @darkspirit510 - There was a lot of work done here, but might have been overcome by events. Do you want finish this one? Or start over and do this in smaller chunks?

@darkspirit510 - Nudge :-)

darkspirit510 commented:
Since this is way too old, I'll redo it in smaller chunks 😅
